carbon-design-system/carbon
# fix(combobox): controlled combobox onchange

Closes https://github.com/carbon-design-system/carbon/issues/17847

In cases where a selection is made to a fully controlled combobox component, and the selectedItem prop updates on this selection, such as in cases where this state is managed externally (probably just about all cases where `selectedItem` is being used), the onChange was firing twice. This PR adds a check to see if the selectedItem prop was changed externally before calling the onChange in the useEffect pertaining to controlled comboboxes. It also ensures that onChange is not fired for controlled comboboxes when a selection is made that is the same as the current selection.

* updated useEffect hook for controlled combobox so onChange doesn't fire twice when selection is made and selectedItem prop changes
* updated onSelectedItemChange to not fire onChange if selectedItem is unchanged
* unit tests

#### Testing / Reviewing

* add log to [onChange in fullyControlled combobox in ComboBox.stories.js](https://github.com/Neues/carbon/blob/ef9f8b0418431d885f464effeff4ab98bbea78e8/packages/react/src/components/ComboBox/ComboBox.stories.js#L257)
* select different options to see onChange fires only once
* select the same option to see that onChange does not fire
